### PR TITLE
ghc: Update to version 9.4.2; cabal: Update to version 3.8.1.0; Bootstrapping support

### PR DESCRIPTION
**Pull request by essandess**
This is a major upgrade of `ghc` and `cabal` to provide support for the arm architecture and (`stack`-independent) bootstrapping of Haskell world binaries with new `ghc-prebuilt` and `cabal-prebuilt` ports.

### ghc: Update to version 9.4.2    
* Update to version 9.4.2
* Refactor to use new hadrian bootstrapping
* hadrian port submission
* ghc-prebuilt subport for bootstrapping
* integration with new cabal-prebuilt subport
* integration with revised PG haskell_cabal-1.0.tcl

### cabal: Update to version 3.8.1.0
* Update to version 3.8.1.0
* Add arm64
* Refactor bootstrap approach for ghc-9.4.2
* Update haskell_cabal-1.0.tcl
* Add cabal-prebuilt subport for bootstrapping
* Integration with new ghc-prebuilt subport

#### Description

Note: The CI will not build this PR because it necessarily uses a new stack-based port.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 12.5 21G72 x86_64
Xcode 13.4.1 13F100

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: lang/ghc/Portfile
Comment (reneeotten) [inline lang/ghc/Portfile:158]: Python 3.10 is the current MacPorts' default, can that be used?

File: lang/ghc/Portfile
Comment (essandess) [reply lang/ghc/Portfile:198]: This comment and `use_parallel_build no` was valid for the previous `make`-based build. I'm not sure if it's necessary for the new `hadrian`-based build, and I've observed it builds fine with parallel builds.

I suggest keeping it commented out and adding to back in if the `hadrian` parallel build throttles the buildbots.

File: lang/ghc/Portfile
Comment (essandess) [reply lang/ghc/Portfile:158]: Yes, I'll add that.

**Comment by reneeotten** [comment]
the failure in the CI log appears to be:
```
2022-08-19T13:15:41.3041790Z Non optional builder 'sphinx-build' is not specified in 'hadrian/cfg/system.config' file.
2022-08-19T13:15:41.3042250Z Build failed.
```

so you probably need to patch that to be `sphinx-build-3.10` instead.

**Comment by essandess** [comment]
> the failure in the CI log appears to be:
> 
> ```
> 2022-08-19T13:15:41.3041790Z Non optional builder 'sphinx-build' is not specified in 'hadrian/cfg/system.config' file.
> 2022-08-19T13:15:41.3042250Z Build failed.
> ```
> 
> so you probably need to patch that to be `sphinx-build-3.10` instead.

Thanks. I created a symlink for `sphinx-build`, as in my existing `port select` settings that I overlooked.

That file is generated by `autoconf` so there's nothing to patch.

**Comment by essandess** [comment]
I see this build issue on ARM. I suggest holding off merging until this can be resolved.
```sh
:info:build Command line: _build/stage1/bin/haddock --verbosity=0 -B_build/stage1/lib --lib=_build/stage1/lib --odir=_build/doc/html/libraries/ghc-prim-0.9.0 --dump-interface=_build/doc/html/libraries/ghc-prim-0.9.0/ghc-prim.haddock --html --hyperlinked-source --hoogle --quickjump '--title=ghc-prim-0.9.0: GHC primitives' --prologue=_build/doc/html/libraries/ghc-prim-0.9.0/haddock-prologue.txt --optghc=-D__HADDOCK_VERSION__=2270 --optghc=-hisuf --optghc=dyn_hi --optghc=-osuf --optghc=dyn_o --optghc=-hcsuf --optghc=dyn_hc --optghc=-fPIC --optghc=-dynamic --optghc=-hide-all-packages --optghc=-no-user-package-db '--optghc=-package-env -' '--optghc=-this-unit-id ghc-prim-0.9.0' '--optghc=-package-id rts-1.0.2' --optghc=-i --optghc=-i/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/src/ghc-9.4.1/_build/stage1/libraries/ghc-prim/build --optghc=-i/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/src/ghc-9.4.1/_build/stage1/libraries/ghc-prim/build/autogen --optghc=-i/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/src/ghc-9.4.1/libraries/ghc-prim --optghc=-Irts/include --optghc=-I_build/stage1/libraries/ghc-prim/build --optghc=-I/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/src/ghc-9.4.1/_build/stage1/lib/aarch64-osx-ghc-9.4.1/rts-1.0.2/include --optghc=-optP-include --optghc=-optP_build/stage1/libraries/ghc-prim/build/autogen/cabal_macros.h --optghc=-optc--target=arm64-apple-darwin --optghc=-outputdir --optghc=_build/stage1/libraries/ghc-prim/build --optghc=-this-unit-id --optghc=ghc-prim --optghc=-XHaskell2010 --optghc=-no-global-package-db --optghc=-package-db=/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/src/ghc-9.4.1/_build/stage1/lib/package.conf.d --optghc=-ghcversion-file=rts/include/ghcversion.h --optghc=-ghcversion-file=rts/include/ghcversion.h --optghc=-Wno-deprecated-flags --optghc=-Wno-trustworthy-s
afe libraries/ghc-prim/GHC/CString.hs libraries/ghc-prim/GHC/Classes.hs libraries/ghc-prim/GHC/Debug.hs libraries/ghc-prim/GHC/Magic.hs libraries/ghc-prim/GHC/Magic/Dict.hs libraries/ghc-prim/GHC/Prim/Exception.hs libraries/ghc-prim/GHC/Prim/Ext.hs libraries/ghc-prim/GHC/Prim/Panic.hs libraries/ghc-prim/GHC/Prim/PtrEq.hs _build/stage1/libraries/ghc-prim/build/GHC/PrimopWrappers.hs libraries/ghc-prim/GHC/Tuple.hs libraries/ghc-prim/GHC/Types.hs _build/stage1/libraries/ghc-prim/build/GHC/Prim.hs +RTS -t_build/stage1/haddock-timing-files/ghc-prim.t --machine-readable -RTS
:info:build ===> Command failed with error code: -6
:info:build dyld[89207]: symbol not found in flat namespace (_ffi_call)
:info:build Command failed
:info:build Build failed.
:info:build Command failed: export GHC=/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/bootstrap/lib/ghc-9.4.1/bin/ghc PATH=/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/bootstrap/lib/ghc-9.4.1/bin:/opt/local/var/macports/build/_opt_local_ports_lang_ghc/ghc/work/bin:/opt/local/bin:/opt/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin MACPORTS_LEGACY_SUPPORT_DISABLED=1 CC=/usr/bin/clang CXX=/usr/bin/clang++; hadrian -j8 docs
:info:build Exit code: 1
:error:build Failed to build ghc: command execution failed
```

**Comment by essandess** [comment]
And now there’s an issue with 9.4.2: https://gitlab.haskell.org/ghc/ghc/-/issues/22118. I expect this update will be a while and must wait for 9.4.3.

**Comment by kencu** [comment]
the cabal Portfile won't parse if ghc is not already installed:

```
set ghc_version [exec sh -c "ghc --version | sed -E 's|^.+version (\[\[:digit:]])|\\1|'"]
```

```
$ cd lang/cabal
$ port info
Can't map the URL 'file://.' to a port description file ("sh: ghc: command not found").
Please verify that the directory and portfile syntax are correct.
To use the current port, you must be in a port's directory.
```

**Comment by kencu** [comment]
oh, you've changed it since I downloaded it -- apologies. let me recheck again.

**Comment by kencu** [comment]
doesn't seem to like ccache much:

```
checking whether CC supports -no-pie... no
checking for extra options to pass gcc when compiling via C... 
checking Setting up CFLAGS, LDFLAGS, IGNORE_LINKER_LD_FLAGS and CPPFLAGS... done
checking Setting up CONF_CC_OPTS_STAGE0, CONF_GCC_LINKER_OPTS_STAGE0, CONF_LD_LINKER_OPTS_STAGE0 and CONF_CPP_OPTS_STAGE0... done
checking Setting up CONF_CC_OPTS_STAGE1, CONF_GCC_LINKER_OPTS_STAGE1, CONF_LD_LINKER_OPTS_STAGE1 and CONF_CPP_OPTS_STAGE1... done
checking Setting up CONF_CC_OPTS_STAGE2, CONF_GCC_LINKER_OPTS_STAGE2, CONF_LD_LINKER_OPTS_STAGE2 and CONF_CPP_OPTS_STAGE2... done
checking C++ standard library flavour... ./configure: line 8123: ccache /usr/bin/clang++: No such file or directory
configure: error: Failed to compile test program
Command failed:  cd "/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_lang_ghc/ghc-prebuilt/work/ghc-9.4.2" && ./configure --prefix=/opt/local 
```

trying again with ccache disabled...

yep, likes that better:

```
checking Setting up CFLAGS, LDFLAGS, IGNORE_LINKER_LD_FLAGS and CPPFLAGS... done
checking Setting up CONF_CC_OPTS_STAGE0, CONF_GCC_LINKER_OPTS_STAGE0, CONF_LD_LINKER_OPTS_STAGE0 and CONF_CPP_OPTS_STAGE0... done
checking Setting up CONF_CC_OPTS_STAGE1, CONF_GCC_LINKER_OPTS_STAGE1, CONF_LD_LINKER_OPTS_STAGE1 and CONF_CPP_OPTS_STAGE1... done
checking Setting up CONF_CC_OPTS_STAGE2, CONF_GCC_LINKER_OPTS_STAGE2, CONF_LD_LINKER_OPTS_STAGE2 and CONF_CPP_OPTS_STAGE2... done
checking C++ standard library flavour... libc++
checking for linkage against 'c++ c++abi'... success
```

**Comment by essandess** [comment]
> doesn't seem to like ccache much:

I do not observe this issue. My general recommendation is standard practice to merge a working PR with the latest OS versions and subsequently modify for specific legacy issues.

**Comment by essandess** [comment]
This is ready to merge to the best of my knowledge.

It will provide a non-`stack` based approach to build native arm binaries of Haskell projects. The new port `hadrian` (necessary to bootstrap `ghc`) provides a self-contained example.

I confirm that it builds and runs on these systems:

> macOS 12.5.1 21G83 x86_64
> Xcode 13.4.1 13F100

> macOS 12.5.1 21G83 arm64
> Xcode 13.4.1 13F100

**Comment by essandess** [comment]
These ports, also necessary to bootstrap`ghc`, illustrate the new bootstrapped build chain:
* #15998
* #16000
* #16001

**Comment by kencu** [comment]
> > doesn't seem to like ccache much:
> 
> I do not observe this issue. My general recommendation is standard practice to merge a working PR with the latest OS versions and subsequently modify for specific legacy issues.

Interesting that ccache works for you but not for me.

Not really a legacy issue, that, but we'll see what happens in the wild once it goes out there.

**Comment by mascguy** [comment]
> These ports, also necessary to bootstrap`ghc`, illustrate the new bootstrapped build chain:
> * [alex: Update to version 3.2.7.1, native build with bootstrapped cabal #15998](https://github.com/macports/macports-ports/pull/15998)
> * [happy: native build with bootstrapped cabal #16000](https://github.com/macports/macports-ports/pull/16000)
> * [hscolour: native build with bootstrapped cabal #16001](https://github.com/macports/macports-ports/pull/16001)

Just to confirm, do those need to be merged first? Or does this one?

Going forward, please ensure the necessary merge order is crystal-clear... and present in the PR description itself.

**Comment by essandess** [comment]
> > These ports, also necessary to bootstrap`ghc`, illustrate the new bootstrapped build chain:
> > 
> > * [alex: Update to version 3.2.7.1, native build with bootstrapped cabal #15998](https://github.com/macports/macports-ports/pull/15998)
> > * [happy: native build with bootstrapped cabal #16000](https://github.com/macports/macports-ports/pull/16000)
> > * [hscolour: native build with bootstrapped cabal #16001](https://github.com/macports/macports-ports/pull/16001)
> 
> Just to confirm, do those need to be merged first? Or does this one?
> 
> Going forward, please ensure the necessary merge order is crystal-clear... and present in the PR description itself.

This PR is necessary to merge first.

To explain so you and other know the context:

It is necessary to bootstrap Haskell world. In MacPorts prior to this PR, Haskell is bootstrapped via a combination of `stack`-based tools and `ghc` self-bootstrapping. This PR allows `stack`-independent bootstrapping completely within `ghc` and `cabal` builds. Those other PRs use this one to provide a bootstrapped Haskell world completely independent of `stack`. Furthermore, this new bootstrapped Haskell world provides native builds of all tools, unlike `x86_64`-only `stack`, and is therefore preferred. Of course, `stack`-based builds are still an option.

**Review by mascguy — APPROVED** [review]

**Comment by essandess** [comment]
@kencu I know that you and others are interested in getting Haskell world onto legacy systems. I can't promise that this new option will be easier than it was with `stack`, but at least it offers another path.

I imagine one approach for legacy systems would be to add some logic to the `ghc-prebuilt` and possibly `cabal-prebuilt` ports to get a compiler/builder up and running, then compiling code would work.

Also note that several libraries need to be updated to support `ghc-9.4.2`-based builds, and that upstream projects like `pandoc` don't yet build with this toolchain. See e.g. https://github.com/jgm/pandoc/issues/8276. As stuff gets updated, we can migrate to MacPorts-native `ghc`/`cabal` builds for these ports.

**Comment by essandess** [comment]
> These ports, also necessary to bootstrap`ghc`, illustrate the new bootstrapped build chain:
> 
> * [alex: Update to version 3.2.7.1, native build with bootstrapped cabal #15998](https://github.com/macports/macports-ports/pull/15998)
> * [happy: native build with bootstrapped cabal #16000](https://github.com/macports/macports-ports/pull/16000)
> * [hscolour: native build with bootstrapped cabal #16001](https://github.com/macports/macports-ports/pull/16001)

@mascguy These all build successfully with the new bootstrapped Haskell world. Once merged, we'll have `stack`-free bootstrapped Haskell world within MacPorts.

**Comment by essandess** [comment]
@mascguy We don't have a comprehensive build on this since #16089 fixed dependencies that presumably break these builds: https://ports.macports.org/port/ghc/details/

I have no insight into the buildbot system—may I impose upon you to launch a build both to have a MacPorts build available for fast download and to see how far back this new Haskell toolchain works?

**Comment by mascguy** [comment]
> @mascguy We don't have a comprehensive build on this since #16089 fixed dependencies that presumably break these builds: https://ports.macports.org/port/ghc/details/
> 
> I have no insight into the buildbot system—may I impose upon you to launch a build both to have a MacPorts build available for fast download and to see how far back this new Haskell toolchain works?

Everyone has the same access to the buildbot history, etc, via: https://ports.macports.org

Search for the port(s) you're interested in, and that will provide the current build status. Or switch to the "Builds" tab, to see the complete history.

**Comment by essandess** [comment]
Thanks. I see the history, but I believe the build should be run again. Perhaps I should have added those `cabal ` dependencies as lib? They’re necessary both to build and run. Would that kick of a new `ghc` build?

**Comment by mascguy** [comment]
> Thanks. I see the history, but I believe the build should be run again. Perhaps I should have added those `cabal ` dependencies as lib? They’re necessary both to build and run. Would that kick of a new `ghc` build?

I'm happy to help, but need a bit more detail, re: what needs to be rebuilt, etc. Can you elaborate?

**Comment by essandess** [comment]
Thanks. Likely superseded by #16115.

**Pull request merged**
